### PR TITLE
Refactor language-server package

### PR DESCRIPTION
**Pull request by Princesseuh**
## Changes

This is a refactor of the language-server package. Do not be fooled by the large amount of changes as this PR also finally apply the formatting changes added in https://github.com/withastro/language-tools/pull/160 which makes for a large portion of the changes.

A lot of the changes especially in the TypeScript plugin are inspired by updates that were made to the Svelte language-server since we forked off them, notably, a lot of the changes they've done in the last year or so are performance related which by proxy also happens to fix a few cache issues we've been having. Huge credits to the Svelte team for their work, always very inspiring. Some changes were also inspired by [Volar](https://github.com/johnsoncodehk/volar), another very impressive language-server codebase to look at

Certain parts were not touched much (if at all) because they'll be subject to planned changes in the near future, such as the AstroPlugin and the Astro Check command. Work on those will be done in dedicated PRs

No new features were added apart from a completion for the `:global` selector in CSS because I noticed that Svelte had it and we didn't. It felt pretty inoffensive to add since we already supported adding completions for selectors

Due to architectural changes, this fixes https://github.com/withastro/language-tools/issues/202
An initial draft of an internal documentation was added, which close https://github.com/withastro/language-tools/issues/201, further work on it will be in dedicated issues to specific subjects
The new VS Code test added fix https://github.com/withastro/language-tools/issues/210
I believe that was already fixed before but even then, this fixes https://github.com/withastro/language-tools/issues/106 (and also does it for non-astro components and js/ts files)

Thanks to the changes in the entrypoints, this setup the way for https://github.com/withastro/language-tools/issues/180 (but we are not quite there yet)

## Testing

New tests setups were added for both the extension itself and the language-server, instead of using the `astro check` command, we now directly test the language-server itself using a similar setup to the one we use in the main repo (Mocha + Chai).

Unlike the previous tests and the main repo, I wrote the tests using TypeScript because it is in my opinion really hard to work with the VS Code packages without types due to the many many different types with similar shapes and names

Additional testing was done manually using various Astro projects

## Docs

Internal documentation was added in the `docs` folder, no docs should be needed on user side

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -52,5 +52,10 @@ jobs:
       - name: Build the packages
         run: yarn build
 
+      - name: Test (Linux)
+        run: xvfb-run -a yarn test
```
Comment (Princesseuh) [inline .github/workflows/ci.yml:56]: This is necessary for the VS Code tests, see [this page](https://code.visualstudio.com/api/working-with-extensions/continuous-integration#github-actions)

File: packages/language-server/src/node.ts [hunk 2]
```diff
@@ -0,0 +1,6 @@
+import * as vscode from 'vscode-languageserver/node';
+import { startLanguageServer } from './server';
+
+const connection = vscode.createConnection(vscode.ProposedFeatures.all);
+
+startLanguageServer(connection);
```
Comment (matthewp) [inline packages/language-server/src/node.ts:6]: Just curious, is having this in its own file setting up to better support non-Node environments?

**Review by matthewp — APPROVED** [review]
Looks great!

File: packages/language-server/test/check/diagnostics.test.ts [hunk 3]
```diff
@@ -18,7 +18,7 @@ describe('Astro Check', async () => {
 		const [{ diagnostics }] = await checker.getDiagnostics();
 
 		expect(diagnostics.length, 'Expected diagnostics to be empty').to.equal(0);
-	});
+	}).timeout(10000);
```
Comment (Princesseuh) [inline packages/language-server/test/check/diagnostics.test.ts:21]: I'm not sure why but the first test in this file is taking a very long time. The tests in this file are kinda faulty in general and will be removed in the future (we don't need to test through AstroCheck anymore)

**Pull request merged**